### PR TITLE
feat: `dfxvm --list` now supports listing the remote available dfx versions.

### PR DESCRIPTION
**Pull request by vincent-dfinity**
# Description

`dfxvm --list` now supports listing the available dfx versions.
  - `--available`: List the available versions.
  - `--limit`:The maximum number of available versions to list in reverse chronological order, with default value `10`.

Fixes # (issue)

[SDK-1883](https://dfinity.atlassian.net/browse/SDK-1883)

# How Has This Been Tested?

1. Manually tested
2. Test added.

# Checklist:

- [x] I have edited the CHANGELOG accordingly.
- [x] I have made corresponding changes to the documentation in docs/cli-reference.

[SDK-1883]: https://dfinity.atlassian.net/browse/SDK-1883?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/dfxvm/list.rs [hunk 1]
```diff
@@ -1,20 +1,42 @@
...
+            println!("No versions available.");
         } else {
-            ""
-        };
-        println!("{}{}", version, default_indicator);
+            println!("Available versions:");
+            let count = std::cmp::min(opts.number, manifest.versions.len());
+            let versions = manifest.versions.iter().rev().take(count);
+            for version in versions {
+                println!("  {}", version);
```
[the remark this answers is not in the document]
Comment (vincent-dfinity) [reply src/dfxvm/list.rs]: Fixed at https://github.com/dfinity/dfxvm/pull/89/commits/7b2e0d67c39ae044dc310841cf0d97e5cf8b1cd6

File: src/dfxvm/list.rs [hunk 2]
```diff
@@ -1,20 +1,42 @@
...
+        info!("fetching {url}");
+        let manifest = fetch_json::<Manifest>(&url).await?;
+
+        if manifest.versions.is_empty() {
+            println!("No versions available.");
         } else {
-            ""
-        };
-        println!("{}{}", version, default_indicator);
+            println!("Available versions:");
```
[the remark this answers is not in the document]
Comment (vincent-dfinity) [reply src/dfxvm/list.rs]: Fixed at https://github.com/dfinity/dfxvm/pull/89/commits/7b2e0d67c39ae044dc310841cf0d97e5cf8b1cd6

File: src/dfxvm/list.rs [hunk 3]
```diff
@@ -1,20 +1,42 @@
...
+        let url = Url::parse(&settings.manifest_url())?;
 
-    for version in installed_versions(locations)? {
-        let default_indicator = if default_version.as_ref() == Some(&version) {
-            " (default)"
+        info!("fetching {url}");
+        let manifest = fetch_json::<Manifest>(&url).await?;
+
+        if manifest.versions.is_empty() {
+            println!("No versions available.");
```
[the remark this answers is not in the document]
Comment (vincent-dfinity) [reply src/dfxvm/list.rs]: Fixed at https://github.com/dfinity/dfxvm/pull/89/commits/7b2e0d67c39ae044dc310841cf0d97e5cf8b1cd6

File: src/dfxvm/cli.rs [hunk 4]
```diff
@@ -42,9 +42,17 @@ pub struct DefaultOpts {
     version: Option<Version>,
 }
 
-/// List installed versions of dfx
+/// List installed or remote available versions of dfx
```
Comment (ericswanson-dfinity) [inline src/dfxvm/cli.rs:45]:
```suggestion
/// List installed or available versions of dfx
```

File: src/dfxvm/cli.rs [hunk 5]
```diff
@@ -42,9 +42,17 @@ pub struct DefaultOpts {
...
 #[derive(Parser)]
-pub struct ListOpts {}
+pub struct ListOpts {
+    /// Specifies to list the remote available versions.
```
Comment (ericswanson-dfinity) [inline src/dfxvm/cli.rs:48]:
```suggestion
    /// Specifies to list the available versions.
```

File: src/dfxvm/cli.rs [hunk 6]
```diff
@@ -42,9 +42,17 @@ pub struct DefaultOpts {
...
+    #[arg(long)]
+    pub available: bool,
+
+    /// Specifies the number of remote versions to be listed, in reverse chronological order.
```
Comment (ericswanson-dfinity) [inline src/dfxvm/cli.rs:52]:
```suggestion
    /// Specifies the number of available versions to be listed, in reverse chronological order.
```

File: src/dfxvm/list.rs [hunk 7]
```diff
@@ -1,20 +1,39 @@
...
+        info!("fetching {url}");
+        let manifest = fetch_json::<Manifest>(&url).await?;
+
+        if !manifest.versions.is_empty() {
+            let count = std::cmp::min(opts.limit, manifest.versions.len());
+            let versions = manifest.versions.iter().rev().take(count);
+            for version in versions {
+                println!("{}", version);
+            }
+        }
```
Comment (ericswanson-dfinity) [inline src/dfxvm/list.rs:25]: This should work without the empty check
```suggestion
        let count = std::cmp::min(opts.limit, manifest.versions.len());
        let versions = manifest.versions.iter().rev().take(count);
        for version in versions {
            println!("{}", version);
        }
```

File: src/error/dfxvm.rs [hunk 8]
```diff
@@ -46,9 +46,15 @@ pub enum Error {
 
 #[derive(Error, Debug)]
 pub enum ListError {
+    #[error("failed to fetch versions")]
```
Comment (ericswanson-dfinity) [inline src/error/dfxvm.rs:49]:
```suggestion
    #[error(transparent)]
```

File: tests/suite/dfxvm/list.rs [hunk 9]
```diff
@@ -87,3 +89,21 @@ fn ignores_non_versions() {
...
+    home_dir
+        .dfxvm()
+        .arg("list")
+        .arg("--available")
+        .assert()
+        .success()
+        .stderr(is_match("info: fetching http://.*/manifest.json").unwrap())
+        .stdout(contains("0.5.2"))
+        .stdout(contains("0.5.0"));
+}
```
Comment (ericswanson-dfinity) [inline tests/suite/dfxvm/list.rs:109]: Please add a test that passes `--limit 1` and make sure the output only shows the expected version

**Review by ericswanson-dfinity — CHANGES_REQUESTED** [review]

File: src/error/dfxvm.rs (hunk 8 above)
Comment (vincent-dfinity) [reply src/error/dfxvm.rs:49]: I followed [UpdateError](https://github.com/dfinity/dfxvm/blob/main/src/error/dfxvm.rs#L73) which also fetches the manifest URL, should we keep them consistent?

File: src/dfxvm/cli.rs (hunk 4 above)
Comment (vincent-dfinity) [reply src/dfxvm/cli.rs]: Fixed at https://github.com/dfinity/dfxvm/pull/89/commits/6bf6a71f54c29ded886d079bb1ee2d95b3756d54

File: src/dfxvm/cli.rs (hunk 5 above)
Comment (vincent-dfinity) [reply src/dfxvm/cli.rs]: Fixed at https://github.com/dfinity/dfxvm/pull/89/commits/6bf6a71f54c29ded886d079bb1ee2d95b3756d54

File: src/dfxvm/cli.rs (hunk 6 above)
Comment (vincent-dfinity) [reply src/dfxvm/cli.rs]: Fixed at https://github.com/dfinity/dfxvm/pull/89/commits/6bf6a71f54c29ded886d079bb1ee2d95b3756d54

File: src/dfxvm/list.rs (hunk 7 above)
Comment (vincent-dfinity) [reply src/dfxvm/list.rs]: Fixed at https://github.com/dfinity/dfxvm/pull/89/commits/6bf6a71f54c29ded886d079bb1ee2d95b3756d54

File: tests/suite/dfxvm/list.rs (hunk 9 above)
Comment (vincent-dfinity) [reply tests/suite/dfxvm/list.rs:129]: Fixed at https://github.com/dfinity/dfxvm/pull/89/commits/6bf6a71f54c29ded886d079bb1ee2d95b3756d54

File: src/error/dfxvm.rs (hunk 8 above)
Comment (vincent-dfinity) [reply src/error/dfxvm.rs]: I updated it anyway at https://github.com/dfinity/dfxvm/pull/89/commits/6bf6a71f54c29ded886d079bb1ee2d95b3756d54

File: src/dfxvm/cli.rs [hunk 10]
```diff
@@ -42,9 +42,17 @@ pub struct DefaultOpts {
     version: Option<Version>,
 }
 
-/// List installed versions of dfx
+/// List installed or available versions of dfx
 #[derive(Parser)]
-pub struct ListOpts {}
+pub struct ListOpts {
+    /// Specifies to list the available versions.
```
Comment (ericswanson-dfinity) [inline src/dfxvm/cli.rs:48]:
```suggestion
    /// List the available versions.
```

File: src/dfxvm/cli.rs [hunk 11]
```diff
@@ -42,9 +42,17 @@ pub struct DefaultOpts {
...
+    #[arg(long)]
+    pub available: bool,
+
+    /// Specifies the number of available versions to be listed, in reverse chronological order.
```
Comment (ericswanson-dfinity) [inline src/dfxvm/cli.rs:52]:
```suggestion
    /// The maximum number of available versions to list, in reverse chronological order.
```

**Review by ericswanson-dfinity — CHANGES_REQUESTED** [review]
Please update https://github.com/dfinity/dfxvm/blob/main/docs/cli-reference/dfxvm/dfxvm-list.mdx

**Review by ericswanson-dfinity — APPROVED** [review]

**Pull request merged**